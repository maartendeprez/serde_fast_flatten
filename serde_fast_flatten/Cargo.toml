[package]
name = "serde_fast_flatten"
description = "A faster `flatten` implementation for serde"
keywords = ["serde", "serialization", "flatten"]
categories = ["encoding"]

edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
version.workspace = true
# rust-version.workspace = true

[features]
default = []
derive = ["dep:serde_fast_flatten_derive"]

[dependencies]
serde = { version = "1.0.217", features = ["derive"] }
serde_fast_flatten_derive = { version = "0.1.0", optional = true, path = "../serde_fast_flatten_derive" }

[dev-dependencies]
serde_fast_flatten_derive = { version = "0.1.0", path = "../serde_fast_flatten_derive" }
bincode = "1.3.3"
bitcode = { version = "0.6.4", features = ["serde"] }
criterion = "0.5.1"
insta = "1.42.1"
serde_cbor = "0.11.2"
serde_json = "1.0.138"
serde_yaml = "0.9.34"

[[bench]]
name = "benchmark"
harness = false
