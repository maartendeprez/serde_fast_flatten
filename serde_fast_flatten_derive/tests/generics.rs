use insta::assert_snapshot;
use serde_fast_flatten::{DeserializeFields, SerializeFields};

#[derive(SerializeFields, DeserializeFields, PartialEq, Eq, Debug)]
struct Paged<T, U> {
    items: Vec<T>,
    #[serde(flatten)]
    params: PageParams<U>,
}

#[derive(SerializeFields, DeserializeFields, PartialEq, Eq, Debug)]
struct PageParams<T> {
    start: T,
    limit: T,
}

#[test]
fn generics() {
    let x = Paged::<String, u64> {
        items: Vec::from_iter([String::from("item 1")]),
        params: PageParams {
            start: 0,
            limit: 10,
        },
    };
    let s = serde_json::to_string_pretty(&x).unwrap();
    let y = serde_json::from_str::<Paged<String, u64>>(&s).unwrap();
    assert_eq!(x, y);
    assert_snapshot!(s);
}
