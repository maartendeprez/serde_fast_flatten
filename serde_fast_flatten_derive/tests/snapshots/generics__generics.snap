---
source: serde_fast_flatten_derive/tests/generics.rs
expression: s
---
{
  "items": [
    "item 1"
  ],
  "start": 0,
  "limit": 10
}
